### ISSUE TITLE
iSEEhex

### ISSUE THREAD

**Issue by kevinrue** [Bioconductor/Contributions#2686]
Tagging co-authors: @federicomarini, @LTLA,  @csoneson

Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/iSEE/iSEEhex

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand Bioconductor [Package Naming Policy][9] and acknowledge
  Bioconductor may retain use of package name.

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.
  
- [x] I am familiar with the [Bioconductor code of conduct][7] and 
  agree to abide by it.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For questions/help about the submission process, including questions about
the output of the automatic reports generated by the SPB (Single Package
Builder), please use the #package-submission channel of our Community Slack.
Follow the link on the home page of the [Bioconductor website][8] to sign up.

[1]: https://contributions.bioconductor.org/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/
[7]: https://bioconductor.org/about/code-of-conduct/
[8]: https://bioconductor.org/
[9]: https://bioconductor.org/developers/package-submission/#naming

**Comment by kevinrue** [comment]
Side note: this is a rather late follow up on the review of the iSEEu package, see https://github.com/Bioconductor/Contributions/issues/1425#issuecomment-601929971

The idea here is to split out functionality related to plot panels that use the `hexbin` package for summarising data into hexagonal bins.
This package contains a carbon copy of code in `iSEEu`, that will be removed from iSEEu when this package is accepted. `iSEEu` would then re-export panels from `iSEEhex`, just like `tidyverse` attaches `dplyr` and friends when it is attached itself.

**Comment by kevinrue** [comment]
Hi,

Following the instructions in the latest message above, i set up the remote and tried to push but i get a hook declined error:
```
% git push upstream main
Warning: No xauth data; using fake authentication data for X11 forwarding.
X11 forwarding request failed on channel 0
Enumerating objects: 13, done.
Counting objects: 100% (13/13), done.
Delta compression using up to 16 threads
Compressing objects: 100% (6/6), done.
Writing objects: 100% (7/7), 1.21 KiB | 1.21 MiB/s, done.
Total 7 (delta 4), reused 0 (delta 0), pack-reused 0
remote: FATAL: W refs/heads/main packages/iSEEhex k.rue-albrecht DENIED by fallthru
remote: error: hook declined to update refs/heads/main
To git.bioconductor.org:packages/iSEEhex.git
 ! [remote rejected] main -> main (hook declined)
error: failed to push some refs to 'git.bioconductor.org:packages/iSEEhex.git'
```

I'm not quite sure where the `k.rue-albrecht` comes from in that message above. It doesn't look exactly like any of my identifiers that I'm aware of. Is that something on the Bioconductor side?

Also, the link to the quick tutorial includes the `.git` extension to the remote, as in
```
git@git.bioconductor.org:packages/<YOUR-REPOSITORY-NAME>.git # <- that final '.git'
```
while the message above ends in `iSEEhex`.
I've tried both versions and neither one works.

**Comment by kevinrue** [comment]
Nevermind. I didn't read far enough into the tutorial. I had the textbook issue of having a `main` branch on GitHub that needed extra work for pushing to the `master` branch on BioC. All good now.

**Comment by lshep** [comment]
we are experiencing some issues with the newly added windows builder that is prevent the new build report from being posted. We are working on a remedy.  We appreciate your patience and understanding.

**Comment by LiNk-NY** [comment]
Hi Kevin, @kevinrue 

Thank you for your submission. 
Please see the review below. 

Best,
Marcel

---

# iSEEhex #2686

## DESCRIPTION

* Looks good.

## NAMESPACE

* The `exportMethods` with the `.function` names seem a bit strange. The dot
is usually reserved for unexported functions but it looks like this is
convention in `iSEE`.

## vignettes/

* Consider preserving the aspect ratio (see screenshot) when plotting.
On widescreen, the plots are wider than they are tall.
![iSEEhex_screenshot](https://user-images.githubusercontent.com/4392950/174104537-f5cf91bf-aa79-48da-9615-4b3712f248dc.png)

## R/

* Document slots with the `@slots` tag.
* Optional. Use `@describeIn` to document methods.

## tests/

* Package coverage looks good.

```r
covr::package_coverage()
iSEEhex Coverage: 61.62%
R/ReducedDimensionHexPlot.R: 61.62%
```

**Comment by kevinrue** [comment]
Thanks. All good points. I'll get on them early next week.

**Comment by kevinrue** [comment]
Just to comment on this one ( @LTLA might want to weigh in, too, as he in introduced this practice in `iSEE` )

> The exportMethods with the .function names seem a bit strange. The dot
is usually reserved for unexported functions but it looks like this is
convention in iSEE.

It did feel awkward when we first adopted this practice. However, the logic to me is that our user base is composed of two groups:

- panel developers
- panel users

The users need access to only a subset of functions (mainly the constructor), to create and configure instances of panels and launch the shiny app. The functions without `.` are intended for those users.

Panel developers need access to additional functions (actually, generic methods) to customise the behaviour of existing panels classes or develop new classes of panels.
The functions that start with `.` are intended for those users.

The extra `.` being intended as a visual cue separating those 'semi-internal' functions from the basic usage functions.

With all that said, I don't have any major preference either way, and I wonder whether this naming style really does have any impact on package users with various levels of familiarity with R and Bioconductor package coding styles. 

In summary, for the time being, I would prefer to keep the current status quo for the purpose of this submission, but I'd be open to remove the `.` from all those generics and methods in the future if the argument is made that it does not benefit users and developers in any way.

**Comment by kevinrue** [comment]
On this point:

> Consider preserving the aspect ratio (see screenshot) when plotting.
On widescreen, the plots are wider than they are tall.

I'm confused. I've opened the vignette on two different monitors, stretched the window, and I never have any issue with the figure dimensions. At worst, the figure is right-aligned and not using all 100% of available width, but it keeps the same aspect ratio (tested on Firefox and Safari).

<img width="969" alt="image" src="https://user-images.githubusercontent.com/5769108/174601934-2997706c-26a4-4c62-b2e1-8b2fcab742cc.png">

Which web browser where you using and which HTML file of the vignette did you open when you ran into that issue? 

**EDIT 1** Hold on. You screenshot doesn't look like the screenshot in the vignette. Does that mean that your screenshot comes from a live app that you launched yourself? In which case, it is normal that - under the default settings - panels are stretched to use a certain width that is relative to the width of the monitor.  Clearly, `PanelWidth = 6L` is too wide on your monitor. That is when you must use configure panel width either using the `PanelWidth` setting in the launch script or in the in-app modal window. Examples below (change PanelWidth to a suitable value for you):

```R
initialPanels <- list(
    ReducedDimensionPlot(
        ColorBy = "Feature name", ColorByFeatureName = "Cux2", PanelWidth = 6L),
    ReducedDimensionHexPlot(
        ColorBy = "Feature name", ColorByFeatureName = "Cux2", PanelWidth = 6L,
        BinResolution = 30)
)
app <- iSEE(se = sce, initial = initialPanels)
```

![image](https://user-images.githubusercontent.com/5769108/174604653-09b65b1b-f438-4088-b4c0-aabc492b4a9f.png)

**Comment by kevinrue** [comment]
On the point

> Document slots with the @slots tag.

Here's a preview of the difference between using the `@slot` tag (top) and the current custom section using the `\itemize` LaTeX command.

![image](https://user-images.githubusercontent.com/5769108/174620776-11d78c7d-9dce-4c13-918b-e3cad9934639.png)

I understand that `@slot` is best practice in roxygen.
@LTLA, do you remember any particular reason for the latter so far? Free form documentation I suppose, as we can add a custom statement that redirects to the slots defined in the parent class(es), which we cannot do in the automatic `Slots` section, as far as I am aware.

To be fair, I would hate to re-include all the definitions of slots from the parent classes, and I find the current custom section convenient to describe the new slots accompanied by a statement that redirects to all the parent classes for the relevant other slots.

Thoughts?

**Comment by kevinrue** [comment]
On the point

> Optional. Use @describeIn to document methods.

Similar to the comment above on the `@slot` tag. We are currently using a dedicated section to benefit of the freedom to mix a description of the various methods using `\itemize` while also being able to add some additional comments. For instance:

![image](https://user-images.githubusercontent.com/5769108/174623777-111063a7-ce73-4caf-8d01-7a1a99f850b7.png)

Note the bit that says:

> In the following code snippets, x is an instance of a ReducedDimensionHexPlot class. Refer to the documentation for each method for more details on the remaining arguments.

Using the `@describeIn` tag might make it more convenient to physically relocate the documentation of each method next to the code of the method, while re-combining everything nicely into a list of bullet points (as below). However, it removes the freedom to add the additional comments above.

![image](https://user-images.githubusercontent.com/5769108/174623509-ad3af1dc-1045-4822-b40d-2b75eda47694.png)

I agree that there is an argument for using the roxygen tags to follow what are probably best practices. However, considering this point made me rediscover why we ultimately chose to use custom sections. In this particular case, it allows use to manually categorise families of methods in a way that (we hope) helps users wishing to develop their own iSEE panels.

**Comment by kevinrue** [comment]
@LiNk-NY

My comments above address the individual points of your review. Apologies for the spam in separate messages: each point came with a small discussion and screenshots, and it felt clearer to split the individual arguments, even though some of them turn out to overlap.

In summary, all points made in the review are absolutely valid, and could be addressed, however:

- for the time being, i'd like to keep the current function naming scheme using `.` to separate functions meant for panel developers from those meant for app users.
- the aspect ratio of panels in live apps is something that app users can control at run time (shiny apps set a grid of 12 units wide, and users can dynamically reduce the width of panel down to 1 unit = 1/12 of the window width; at which point, if the panel is still too wide, then you should simply reduce the width of the window that contains the app)
- the `@slot` tag in roxygen is a good suggestion, but I'd like to keep the current custom section format that allows us to redirect to parent classes for inherited slots.
- the `@describeIn` tag is a good suggestion, but I'd like to keep the current custom section format that allows us to group thematically related methods in separate subsections.

The points above essentially argue toward not making any fix toward the points that you raised. Please let me know if you still feel strongly about any of them.

The only other point that you raised was about unit test coverage, which I'm happy to work towards, before or after the package is accepted. (I've got a pretty good track record on test coverage in my other packages)

**Comment by LiNk-NY** [comment]
@kevinrue 

> It did feel awkward when we first adopted this practice. However, the logic to me is that our user base is composed of two groups:
> 
> * panel developers
> * panel users

This sounds like there should one package for each intended audience. It may be too late to change now but it does not look like standard practice. 

WRT `@slots` and `@describeIn`, these are optional. If the free-form documentation is working, that should be good enough though I do like the `Slots` section from `@slots` in the docs.

FWIW, the `setClass` code 
```r
setClass("ReducedDimensionHexPlot", contains="ReducedDimensionPlot", slots=collated)
```
looks a bit obfuscated. Why not simply `slots=c(BinResolution = "numeric")`. It is easier to read. I don't think `collated` is used elsewhere in the code. 

> To be fair, I would hate to re-include all the definitions of slots from the parent classes, and I find the current custom section convenient to describe the new slots accompanied by a statement that redirects to all the parent classes for the relevant other slots.

I think you only need to include documentation for the new slots. You could also use the `@seealso` tag and make additional notes in the documentation page. 

RE: `@describeIn`: The methods certainly can be grouped under a single documentation page so I don't quite follow what you mean. 

I was running the app manually rather than looking at the vignette or the hosted app. Therefore, the scaling is not an issue. 

It looks like it's easier to keep the minor documentation details as-is. 

Thanks for looking into it. I will accept the package. 

Best,
Marcel

**Comment by kevinrue** [comment]
Hi @LiNk-NY 

Thank for the feedback.
As always, this is a very useful discussion to have.

> This sounds like there should one package for each intended audience. It may be too late to change now but it does not look like standard practice

I hadn't seen things in that way yet. Perhaps it is worth indeed looking into something like an `iSEEGenerics` package.

> FWIW, the setClass code

Ideally, I'd agree. And for classes that do no define "too many" slots that is entirely best practices. Maybe for context, have a look at one of the base iSEE classes that define ~ 90% of the slots inherited by more specialised panel classes. That's where and why we pulled out the various slots into separate variables. Not ideal, but those panels are huge classes which take S4 to its limits (in terms of coding style).

https://github.com/iSEE/iSEE/blob/master/R/AllClasses.R#L159

> RE: @describeIn: The methods certainly can be grouped under a single documentation page so I don't quite follow what you mean.

What I mean is that, in that single documentation page, `describeIn` collates the various points into a single list of items. That's very tidy and pretty but what we have at the moment is NOT a single list of bullet points, but multiple lists of bullet points, grouping families of methods by theme (e.g., "GUI-controlling methods", "plot generation methods").

I hope that clarifies your last few points :)

**Comment by LiNk-NY** [comment]
Hi @kevinrue 

> > FWIW, the setClass code
> 
> Ideally, I'd agree. And for classes that do no define "too many" slots that is entirely best practices. Maybe for context, have a look at one of the base iSEE classes that define ~ 90% of the slots inherited by more specialised panel classes. That's where and why we pulled out the various slots into separate variables. Not ideal, but those panels are huge classes which take S4 to its limits (in terms of coding style).
> 
> https://github.com/iSEE/iSEE/blob/master/R/AllClasses.R#L159
> 

Perhaps the data should be encoded as a `data.frame` (with class, slot, and type columns) and stored internally as part of the package. And in some ways, the number slots may indicate that subclasses should take more of a role to encode that data but I don't understand the package thoroughly. 

> > RE: @describeIn: The methods certainly can be grouped under a single documentation page so I don't quite follow what you mean.
> 
> What I mean is that, in that single documentation page, `describeIn` collates the various points into a single list of items. That's very tidy and pretty but what we have at the moment is NOT a single list of bullet points, but multiple lists of bullet points, grouping families of methods by theme (e.g., "GUI-controlling methods", "plot generation methods").
> 
> I hope that clarifies your last few points :)

Ah, I see what you mean. It is also possible to separate the families of methods by documenting them in separate pages.

**Comment by lshep** [comment]
The master branch of your GitHub repository has been added to Bioconductor's git repository.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/kevinrue.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `BiocManager::install("iSEEhex")`. The package 'landing page' will be created at

https://bioconductor.org/packages/iSEEhex

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/
